utilities/scale-json/src/test.rs

```diff
@@ -0,0 +1,173 @@
+#![cfg(test)]
+use codec::{Decode, Encode};
+use scale_decode::DecodeAsType;
+use scale_info::{PortableRegistry, TypeInfo};
+
+use super::*;
+
+#[derive(Debug, Clone, PartialEq, Eq, Encode, Decode, TypeInfo)]
+enum RuntimeEvent {
+	System(SystemEvent),
+	Other(OtherEvent),
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Encode, Decode, TypeInfo)]
+enum SystemEvent {
+	ExtrinsicSuccess { dispatch_info: String },
+	ExtrinsicFailed { dispatch_error: String },
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Encode, Decode, TypeInfo)]
+enum OtherEvent {
+	Unit,
+	Value(u32),
+	Tuple(u32, String, bool),
+	Struct {
+		num_u8: u8,
+		num_u16: u16,
+		num_u32: u32,
+		num_u64: u64,
+		num_u128: u128,
+		num_i8: i8,
+		num_i16: i16,
+		num_i32: i32,
+		num_i64: i64,
+		num_i128: i128,
+	},
+	Array(Vec<u32>),
+	Ethereum(InstanceLike<Ethereum>),
+	Bitcoin(InstanceLike<Bitcoin>),
+	BytesArray([u8; 32]),
+	ByteVec(Vec<u8>),
+}
+
+trait Chain {
+	type Asset;
+}
+
+mod eth {
+	use super::*;
+	#[derive(Debug, Clone, PartialEq, Eq, Encode, Decode, TypeInfo)]
+	pub enum Asset {
+		Eth,
+		Flip,
+		Usdc,
+	}
+}
+
+mod btc {
+	use super::*;
+	#[derive(Debug, Clone, PartialEq, Eq, Encode, Decode, TypeInfo)]
+	pub enum Asset {
+		Btc,
+	}
+}
+
+#[derive(Default, Debug, Clone, PartialEq, Eq, Encode, Decode, TypeInfo)]
+struct Ethereum;
+impl Chain for Ethereum {
+	type Asset = eth::Asset;
+}
+
+#[derive(Default, Debug, Clone, PartialEq, Eq, Encode, Decode, TypeInfo)]
+struct Bitcoin;
+impl Chain for Bitcoin {
+	type Asset = btc::Asset;
+}
+
+#[derive(Default, Debug, Clone, PartialEq, Eq, Encode, Decode, TypeInfo)]
+struct InstanceLike<T: Chain> {
+	inner: T::Asset,
+}
+
+fn make_type_resolver<T: TypeInfo + 'static>() -> (u32, PortableRegistry) {
```

Comment:
just checking my understanding, when using this in a real implementation, we would just have one registry, and all the types would be registered to it? because currently we create a registry for every type separately, one for each test (which is fine for convenience).

Reply:
Yes normally we would get the registry from the metadata and we only need to refresh it on a runtime upgrade. This will become clearer with the upcoming PR #4905